pyro/infer/predictive.py

```diff
@@ -450,3 +450,181 @@ def forward(self, *args, **kwargs):
             guide_log_prob=guide_log_prob,
             model_log_prob=model_log_prob,
         )
+
+
+class MHResampler(torch.nn.Module):
+    r"""
+    Resampler for weighed samples that generates equally weighed samples from the distribution
+    specified by the weighed samples ``sampler``.
+
+    The resampling is based on the Metropolis-Hastings algorithm.
+    Given an initial sample :math:`x` subsequent samples are generated by:
+
+    -   Sampling from the ``guide`` a new sample candidate :math:`x'` with probability :math:`g(x')`.
+    -   Calculate an acceptance probability
+        :math:`A(x', x) = \min\left(1, \frac{P(x')}{P(x)} \frac{g(x)}{g(x')}\right)`
+        with :math:`P` being the ``model``.
+    -   With probability :math:`A(x', x)` accept the new sample candidate :math:`x'`
+        as the next sample, otherwise set the current sample :math:`x` as the next sample.
+
+    The above is the Metropolis-Hastings algorithm with the new sample candidate
+    proposal distribution being equal to the ``guide`` and independent of the
+    current sample such that :math:`g(x')=g(x' \mid x)`.
+
+    :param callable sampler: When called returns :class:`WeighedPredictiveResults`.
+    :param slice source_samples_slice: Select source samples for storage (default is `slice(0)`, i.e. none).
+    :param slice stored_samples_slice: Select output samples for storage (default is `slice(0)`, i.e. none).
+
+    The typical use case of :class:`MHResampler` would be to convert weighed samples
+    generated by :class:`WeighedPredictive` into equally weighed samples from the target distribution.
+
+    Example::
+
+        def model():
+            ...
+
+        def guide():
+            ...
+
+        def conditioned_model():
+            ...
+
+        # Fit guide
+        elbo = Trace_ELBO(num_particles=100, vectorize_particles=True)
+        svi = SVI(conditioned_model, guide, optim.Adam(dict(lr=3.0)), elbo)
+        for i in range(num_svi_steps):
+            svi.step()
+
+        # Create callable that returns weighed samples
+        posterior_predictive = WeighedPredictive(model,
+                                                 guide=beta_guide,
+                                                 num_samples=num_samples,
+                                                 parallel=parallel,
+                                                 return_sites=["_RETURN"])
+
+        prob = 0.95
+
+        weighed_samples = posterior_predictive(model_guide=conditioned_model)
+        # Calculate quantile directly from weighed samples
+        weighed_samples_quantile = weighed_quantile(weighed_samples.samples['_RETURN'],
+                                                    [prob],
+                                                    weighed_samples.log_weights)[0]
+
+        resampler = MHResampler(posterior_predictive)
+        for resampling_count in range(10):
+            resampled_weighed_samples = resampler(model_guide=conditioned_model)
```

Comment:
Do I understand correctly that this discards the first 9 return values and keeps only the 10th? If so, consider adding a comment explaining what the number 10 is, or just naming it `mh_steps = 10`.

Reply:
Yes, you understood correctly, the first 9 sets of samples are discarded in order to allow the resampling process to converge to the true posterior predictive distribution.

Reply:
I've replaced 10 with `num_mh_steps`.